### PR TITLE
Coco/area relational integrity

### PR DESCRIPTION
**Pull request by CocoisBuggy**
Fulfils the feature request in #126 

There's a bit happening here, since I made some performance driven changes to help me with a different PR (#407) I tried to knock out the major test cases but the nature of this particular feature is a little thorny because of how we utilize de-normalized data.

The commit history is a little messy and I am frankly feeling a bit lazy, which is why I have not squashed them. Been moving between machines a bunch and reworked the approach a few times to get speed tuned in.
If someone would like me to squash then I will try and motivate myself to do it.

The key takeaway in terms of changes that will affect other devs is that a new field has been introduced on the area document from which all other ancestry / children are derived. the `parent` field points to the areas direct parent. The pathTokens and ancestors fields have been colocated into a single array of sub documents so that the update propogations can happen all together - not a big deal now, but gives us a sane path to embed  more de-normalized data if we need to.

@vnugent I'd like your particular eyes for these changes.

### CODE REVIEW

**Comment by vnugent** [comment]
@CocoisBuggy can you please explain what problems this PR is trying to solve?  Maybe a diagram or example would help.  Thanks!
Not sure if you notice I created a rough plan for v2 #422 where we can address some of the major issues.  Perhaps we can discuss larger impact changes in v2?

**Comment by CocoisBuggy** [comment]
Sure thing, sorry I suppose I wasn't super clear in my explanation.


The tldr is:
- performance
- code sanity
- collate demoralized fields together
- allow updating of area parents to reorganize areas
- deconvolute data structure

## Who should be affected?

Only developers working on the back end. There is a scary number of changes, but I haven't touched the way in which the data is accessed by the api so there **shouldn't** be a problem - but now that I'm writing that it may be just as well for me to check that the existing tests cover the behavior.

Developers shouldn't have to change their mental model too much, since all the old data Is still available - and is recomputed when it's dependencies update.

## Performance

In my other pr I'm working on some queries that I need for a feature I'm interested in, but performing the lookup for parents of nodes was always a substantial hit when I was looking at my profiler. Partly this is a consequence of storing the ancestors as a comma delimited string which seemed to always be a performance sink whenever I touched it. When using the children array I found a couple of strange hydras when performing deep queries.

After struggling to get the queries down below 100-200ms I decided to look elsewhere for a solution.

## Code Sanity

While I was looking into area update and coming to terms with the existing denormalization strategies I found myself yearning for a single approach to be used so that it could all be debugged and understood together - especially since they're absolutely related code work.

It also made me increasingly perplexed that the children were specified as an array of ObjectId but the ancestors is a string - and a string of uuid values at that.

My solution to future-proof against this problem getting worse over time and also resolve the inconsistencies I decided to put the related data together.

## Area Structure Updating

When I went to implement this feature neither the array of children nor the ancestor string felt like good ways to express which area had which parent. I decided to make these explicitly derived fields, and then simply put a parent field on each area. The difference here is slightly more than philosophical, but without going into too much detail my belief is that this is the easiest way to not get stuck in side effect hell. I can expand more on that if needed, but for the time being it suffices to say that there's no way to accidentally end up with areas belonging to multiple parents, which would have cascading bugs.

So now you need only update an areas parent, and then recompute the paths from the altered node (which is quite speedy).

## conclusion

The pr is more or less worthless if we want to specifically avoid a parent field pattern. It should work well with the v2 milestone, especially for phasing out uuid under the hood. I can do more legwork to make sure it doesn't change the existing api, as it shouldn't be necessary for this pr

**Comment by vnugent** [comment]
@CocoisBuggy Thanks for the large PR.  Can we discuss this after the new year?

**Comment by CocoisBuggy** [comment]
For sure, merry Christmas @vnugent !

**Comment by CocoisBuggy** [comment]
Happy new year fellow climbers! I hope everyone had a good christmas. Down here in SA we are so far out of season we do a nightly prayer for our poor skin.

I thought I'd resolve the merge issues as a nice little start to the year. I'm going to copy the contents of my previous commit as it contains a question I now have:

```
My confidence in this deletion is a little lower than it initially was.
I remember checking the code archeology the first time around and
thinking that we must have been done with importing data (especially
since descriptions have become mutable and I had assumed importing
might result in unintensional overwrites.)

However this change to the code indicates otherwise. Could someone
indicate where in the workflow this code is still in use? If it's still
in use then obviously I actually do need to fix this up for the PR to
make any sense...
```

**Comment by CocoisBuggy** [comment]
#316 
#352
#126
#458
are all partially or fully resolved in this pr, any eyes on this or feedback would be greatly appreciated
